alf/algorithms/functional_particle_vi_algorithm_test.py

```diff
@@ -0,0 +1,169 @@
+
+
+class FuncParVIAlgorithmTest(parameterized.TestCase, alf.test.TestCase):
+    def cov(self, data, rowvar=False):
+        """Estimate a covariance matrix given data.
+
+        Args:
+            data (tensor): A 1-D or 2-D tensor containing multiple observations 
+                of multiple dimentions. Each row of ``mat`` represents a
+                dimension of the observation, and each column a single
+                observation.
+            rowvar (bool): If True, then each row represents a dimension, with
+                observations in the columns. Othewise, each column represents
+                a dimension while the rows contains observations.
+
+        Returns:
+            The covariance matrix
+        """
+        x = data.detach().clone()
+        if x.dim() > 2:
+            raise ValueError('data has more than 2 dimensions')
+        if x.dim() < 2:
+            x = x.view(1, -1)
+        if not rowvar and x.size(0) != 1:
+            x = x.t()
+        fact = 1.0 / (x.size(1) - 1)
+        x -= torch.mean(x, dim=1, keepdim=True)
+        return fact * x.matmul(x.t()).squeeze()
+
+    def assertArrayGreater(self, x, y, eps):
+        self.assertEqual(x.shape, y.shape)
+        self.assertGreater(float(torch.min(x - y)), eps)
+
+    @parameterized.parameters(('gfsf'), ('svgd'), ('gfsf', True),
+                              ('svgd', True))
+    def test_functional_par_vi_algorithm(self,
+                                         par_vi='svgd',
+                                         function_vi=False,
+                                         num_particles=256,
+                                         train_batch_size=10):
+        """
+        The hypernetwork is trained to generate the parameter vector for a linear
+        regressor. The target linear regressor is :math:`y = X\beta + e`, where 
+        :math:`e\sim N(0, I)` is random noise, :math:`X` is the input data matrix, 
+        and :math:`y` is target ouputs. The posterior of :math:`\beta` has a 
+        closed-form :math:`p(\beta|X,y)\sim N((X^TX)^{-1}X^Ty, X^TX)`.
+        For a linear generator with weight W and bias b, and takes standard Gaussian 
+        noise as input, the output follows a Gaussian :math:`N(b, WW^T)`, which should 
+        match the posterior :math:`p(\beta|X,y)` for both svgd and gfsf.
+        
+        """
+        input_size = 3
+        input_spec = TensorSpec((input_size, ), torch.float32)
+        output_dim = 1
+        batch_size = 100
+        inputs = input_spec.randn(outer_dims=(batch_size, ))
+        beta = torch.rand(input_size, output_dim) + 5.
+        print("beta: {}".format(beta))
+        noise = torch.randn(batch_size, output_dim)
+        targets = inputs @ beta + noise
+        true_cov = torch.inverse(
+            inputs.t() @ inputs)  # + torch.eye(input_size))
+        true_mean = true_cov @ inputs.t() @ targets
+        noise_dim = 3
+        algorithm = FuncParVIAlgorithm(
```

Comment:
FuncParVIAlgorithm.train_iter() is not tested. Perhaps use train_iter to do the training here so that everything is tested.

Reply:
Updated.